include/taskolib/Tag.h

```diff
@@ -0,0 +1,128 @@
+/**
+ * \file   Tag.h
+ * \author Lars Fröhlich
+ * \date   Created on July 17, 2024
+ * \brief  Declaration of the Tag class.
+ *
+ * \copyright Copyright 2024 Deutsches Elektronen-Synchrotron (DESY), Hamburg
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 2.1 of the license, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#ifndef TASKOLIB_TAG_H_
+#define TASKOLIB_TAG_H_
+
+#include <iosfwd>
+#include <string>
+
+#include <gul14/string_view.h>
+
+namespace task {
+
+/**
+ * A tag used for categorizing sequences.
+ *
+ * A tag consists of lowercase ASCII letters, digits, and hyphen characters. It must be at
+ * least 1 character long and has a maximum length given by the max_length member
+ * constant. Uppercase letters are automatically converted to lowercase when comparing or
+ * creating tags.
+ */
+class Tag
+{
+public:
+    /// Maximum number of bytes of a tag name.
+    static constexpr std::size_t max_length = 32;
+
+    /// A string containing all of the valid characters for a tag name.
+    static const gul14::string_view valid_characters;
+
+
+    /// Default-construct a tag with the name "-".
+    Tag() : name_{ "-" }
```

Comment:
Why do we have a default constructor?

A default constructed Tag can be used for nothing. It has this fixed content `-`, and you can not change it later on, and it is just unusable for anything (except being replaced immediately by some meaningful tag).

I would suggest removing it (the default constructor), so that only meaningful Tags are possible,

This also needs a removed test and not-creating a vector of a lot (default constructed) Tags that we throw away right away :-D

```diff
diff --git a/include/taskolib/Tag.h b/include/taskolib/Tag.h
index 7f9c2b3..7fa38e0 100644
--- a/include/taskolib/Tag.h
+++ b/include/taskolib/Tag.h
@@ -49,11 +49,6 @@ public:
     /// A string containing all of the valid characters for a tag name.
     static const gul14::string_view valid_characters;
 
-
-    /// Default-construct a tag with the name "-".
-    Tag() : name_{ "-" }
-    {}
-
     /**
      * Construct a tag with the specified name
      *
diff --git a/src/deserialize_sequence.cc b/src/deserialize_sequence.cc
index 26f46e0..e42b38b 100644
--- a/src/deserialize_sequence.cc
+++ b/src/deserialize_sequence.cc
@@ -324,10 +324,10 @@ void load_sequence_parameters(const std::filesystem::path& folder, Sequence& seq
 std::vector<Tag> parse_tags(gul14::string_view str)
 {
     const auto tokens = gul14::tokenize_sv(str);
-
-    std::vector<Tag> tags(tokens.size());
-    std::transform(tokens.begin(), tokens.end(), tags.begin(),
-                   [](gul14::string_view token) { return Tag{ token }; });
+    std::vector<Tag> tags{ };
+    tags.reserve(tokens.size());
+    std::transform( tokens.begin(), tokens.end(), std::back_inserter(tags),
+        [](gul14::string_view token) { return Tag{ token }; } );
     return tags;
 }
 
diff --git a/tests/test_Tag.cc b/tests/test_Tag.cc
index 8e946ca..e896e9e 100644
--- a/tests/test_Tag.cc
+++ b/tests/test_Tag.cc
@@ -31,12 +31,6 @@
 
 using namespace task;
 
-TEST_CASE("Tag: Default constructor", "[Tag]")
-{
-    Tag name;
-    REQUIRE(name.string() == "-");
-}
-
 TEST_CASE("Tag: Constructor from string", "[Tag]")
 {
     Tag a{ "1234" };
```

Reply:
> Why do we have a default constructor?

I added a default constructor because classes without one are difficult to deal with. The `vector` is an example – you cannot set a size at construction, you cannot call `resize()` with default parameters. You cannot even construct an empty `struct` containing a `Tag` unless you give it a constructor or explicitly value-initialize it.

So I would vote "in favor" of the default constructor, but you can overturn it with a "strongly against". :-)

Reply:
No, I'm just "against" and also set "PR approved".

I believe the 'convenience' reasons are not really good reasons, and C++ has tools to accommodate for non-default-constructible classes. Consider:

> The vector is an example – you cannot set a size at construction, you cannot call resize() with default parameters.

That is true. But _wanting_ that as feature is imho the problem here. With resize() you fill it with default constructed (i.e. meaninless) Tags, that need to be removed/replaced afterwards; that needs extra actions. So the correct way is to fill it with meaningful Tags in the first place (see for example code provided in the diff).

> You cannot even construct an empty struct containing a Tag unless you give it a constructor or explicitly value-initialize it.

Well, I also see no problem here. Either specify a default value (in the Struct's ctor) - but that again raises the question if the design is right.

Another way would be to make a Tag changeable, so you can

```c++
Tag t{ };
t.set("foobar");
```

It is only the fact that you can not use a default constructed Tag for anything, not even change it, that makes it (the default constructed Tag) so useless. You can just have it and discard it. And replace with something meaningful.

Whatever, I think we had this discussion already with other classes. Your stance is "but its much more convenient", and my is "but not having a default ctor is more correct and prevents misuse".
Misuse like first creating a possibly big vector with hundredth of default constructed Tag objects and then throwing them all away and create NEW Tags to go into their places. That (creating defaults and replace them immediately) in itself is something to reject in code review I guess; and having no default ctors automatically prevents that.